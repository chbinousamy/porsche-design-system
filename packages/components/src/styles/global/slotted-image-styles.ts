import { TagName } from '@porsche-design-system/shared';
import { addImportantToEachRule } from '../common-styles';
import { Styles } from 'jss';

export const tagNamesWithSlottedImageArray: TagName[] = ['p-table'];

export const getSlottedImageStyles = (tagName: TagName): Styles => ({
  '@global': {
    [`${tagName} img`]: addImportantToEachRule({
      verticalAlign: 'middle',
    }),
  },
});
