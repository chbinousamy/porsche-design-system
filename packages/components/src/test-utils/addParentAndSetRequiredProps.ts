import type { TagName } from '@porsche-design-system/shared';
import { getComponentMeta } from '@porsche-design-system/component-meta';

export const addParentAndSetRequiredProps = (tagName: TagName, component: any): void => {
  const { requiredParent, requiredChild, propsMeta, hasEvent, eventsMeta, requiredNamedSlots } =
    getComponentMeta(tagName);

  if (requiredParent) {
    const parent = document.createElement(requiredParent);
    parent.append(component.host as HTMLElement);
  }

  if (requiredChild) {
    const [childTagName, childAttribute] = requiredChild.split(' ');
    const child = document.createElement(childTagName);
    if (childAttribute) {
      const [childAttributeName, childAttributeValue] = childAttribute.replace(/"/g, '').split('=');
      child[childAttributeName] = childAttributeValue;
    }

    component.host.append(child);
    component[childTagName] = child;
  }

  if (requiredNamedSlots) {
    requiredNamedSlots.forEach(({ slotName, tagName: requiredTagName }) => {
      const child = document.createElement(requiredTagName);
      child.slot = slotName;

      if (requiredTagName.includes('link')) {
        (child as any).href = '#';
      }

      component.host.appendChild(child);
    });
  }

  const requiredProps = Object.entries(propsMeta)
    .filter(([, value]) => value.isRequired)
    .map(([key]) => key);

  if (requiredProps) {
    requiredProps.forEach((prop) => {
      component[prop] = propsMeta[prop].defaultValue ?? 'some value';
    });
  }

  if (hasEvent) {
    Object.keys(eventsMeta).forEach((event) => {
      component[event] = jest.fn();
    });
  }
};
