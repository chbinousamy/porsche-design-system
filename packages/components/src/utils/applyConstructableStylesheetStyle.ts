import { Styles } from 'jss';
import { getMinifiedCss } from '@porsche-design-system/shared';
import { mergeDeep } from './jss';

type ElementsMap = Map<string, ElementMap>;
type ElementMap = Map<Document | ShadowRoot, boolean>;

const elementsMap: ElementsMap = new Map();

const getElementMap = (element: HTMLElement): ElementMap => {
  const tagName = element.tagName;
  if (!elementsMap.has(tagName)) {
    elementsMap.set(tagName, new Map());
  }
  return elementsMap.get(tagName);
};

const hasConstructableStylesheetSupport = ((): boolean => {
  try {
    return typeof new CSSStyleSheet().replaceSync === 'function';
  } catch {
    return false;
  }
})(); // determine it once

export const applyConstructableStylesheetStyles = (
  element: HTMLElement,
  ...getStyles: ((tagName: string) => Styles)[]
): void => {
  if (hasConstructableStylesheetSupport) {
    const documentOrShadowRoot = element.getRootNode() as Document | ShadowRoot;
    const elementMap: ElementMap = getElementMap(element);

    if (!elementMap.has(documentOrShadowRoot)) {
      elementMap.set(documentOrShadowRoot, true);

      const sheet = new CSSStyleSheet();
      sheet.replaceSync(
        getMinifiedCss(mergeDeep(...getStyles.map((getStyle) => getStyle(element.tagName.toLowerCase()))))
      );
      // TODO: for some reason unit test in Docker environment throws TS2339: Property 'push' does not exist on type 'readonly CSSStyleSheet[]'
      /* eslint-disable @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      documentOrShadowRoot.adoptedStyleSheets?.push(sheet);
    }
  }
};
