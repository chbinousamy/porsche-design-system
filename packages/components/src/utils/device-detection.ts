import { hasWindow } from './has-window';

export const isTouchDevice = (): boolean | void => {
  if (!hasWindow) {
    return;
  }
  return !!('ontouchstart' in window || window.navigator.maxTouchPoints > 0);
};
