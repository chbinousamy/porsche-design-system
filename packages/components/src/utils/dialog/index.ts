export { setDialogVisibility, onCancelDialog, onClickDialog } from './dialog';
export { getIntersectionObserverStuck } from './observer';
