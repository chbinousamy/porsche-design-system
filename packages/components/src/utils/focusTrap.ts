export const unpackChildren = (el: HTMLElement | ShadowRoot): HTMLElement[] => {
  return Array.from(el.children, (child: HTMLElement) =>
    child.children ? [child].concat(unpackChildren(child)) : child
  )
    .flat()
    .map((child) => (child.shadowRoot ? [child].concat(unpackChildren(child.shadowRoot)) : child))
    .flat();
};

// TODO: could be extended by audio[controls], video[controls], [contenteditable]:not([contenteditable="false"]) or iframe
export const isFocusableElement = (el: HTMLInputElement): boolean => {
  const { nodeName } = el;
  return (
    ((nodeName === 'INPUT' && el.type !== 'hidden') ||
      nodeName === 'TEXTAREA' ||
      nodeName === 'SELECT' ||
      nodeName === 'BUTTON' ||
      (nodeName === 'A' && !!(el as any).href)) &&
    el.tabIndex >= 0 &&
    !el.disabled
  );
};

export type FirstAndLastFocusableElement = HTMLElement[];
export const getFirstAndLastFocusableElement = (
  host: HTMLElement,
  closeButton?: HTMLElement
): FirstAndLastFocusableElement => {
  const focusableElements = unpackChildren(host).filter(isFocusableElement);

  if (closeButton) {
    return [closeButton, focusableElements.pop() || closeButton];
  }

  return [focusableElements[0], focusableElements.pop()];
};

export let documentKeydownListener: (e: KeyboardEvent) => void;

export const setFocusTrap = (
  host: HTMLElement,
  isOpen: boolean,
  firstFocusableElement: HTMLElement,
  closeBtn?: HTMLElement, // irrelevant for disconnectedCallback
  closeFn?: () => void // irrelevant for disconnectedCallback
): void => {
  let focusableElements: FirstAndLastFocusableElement = [];

  document.removeEventListener('keydown', documentKeydownListener);

  if (isOpen) {
    // the filter removes falsy values: https://stackoverflow.com/a/41346932
    focusableElements = getFirstAndLastFocusableElement(host, closeBtn).filter((x) => x);

    documentKeydownListener = (e: KeyboardEvent): void => {
      const { key, shiftKey } = e;
      const { activeElement } = host.shadowRoot;

      if (key === 'Escape') {
        closeFn();

        return;
      }

      // all other cases respect the natural tab order
      // the cycle itself is accomplished within setFirstAndLastFocusableElementKeydownListener
      if (key !== 'Tab') {
        return;
      }

      // if we don't have any focusableElements we need to prevent Tab here
      if (focusableElements.length === 0) {
        e.preventDefault();

        return;
      }

      // when component is opened initially, the dialog is focused and shift + tab would break out of cycle
      if (shiftKey && activeElement === firstFocusableElement) {
        e.preventDefault();

        focusableElements[1]?.focus();
      }
    };

    document.addEventListener('keydown', documentKeydownListener);
  }

  setFirstAndLastFocusableElementKeydownListener(focusableElements);
};

type KeyboardHandlerTuple = [(e: KeyboardEvent) => void, (e: KeyboardEvent) => void] | [];

/** cache of previous first and last focusable element so we are able to remove them again */
export let FOCUSABLE_ELEMENT_CACHE: FirstAndLastFocusableElement = [];
/** cache of previous event handler pair so we are able to remove them again */
export let KEYDOWN_EVENT_HANDLER_CACHE: KeyboardHandlerTuple = [];

export const setFirstAndLastFocusableElementKeydownListener = (
  focusableElements: FirstAndLastFocusableElement
): void => {
  // remove previous handlers if there are any
  if (FOCUSABLE_ELEMENT_CACHE.length) {
    FOCUSABLE_ELEMENT_CACHE.forEach((el, idx) => el.removeEventListener('keydown', KEYDOWN_EVENT_HANDLER_CACHE[idx]));
  }

  // create, apply and save new handlers for future removal
  if (focusableElements.filter((x) => x).length) {
    FOCUSABLE_ELEMENT_CACHE = [...focusableElements]; // prevent mutation
    KEYDOWN_EVENT_HANDLER_CACHE = focusableElements.map((el, idx) => {
      const handler = (e: KeyboardEvent): void => {
        if (e.key === 'Tab' && ((idx === 0 && e.shiftKey) || (idx === 1 && !e.shiftKey))) {
          e.preventDefault();
          focusableElements[idx === 0 ? 1 : 0].focus();
        }
      };

      el.addEventListener('keydown', handler);

      return handler;
    }) as KeyboardHandlerTuple;
  }
};
