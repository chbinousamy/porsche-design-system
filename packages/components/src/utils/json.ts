export const parseJSONAttribute = <T>(attribute: T | string): T => {
  // Input is object, e.g. { 'aria-label': 'Some label' }
  if (typeof attribute !== 'string') {
    return attribute;
  }

  // Convert single quotes to double quotes except the ones which are escaped by backslash
  let jsonString = attribute
    .replace(/\\'/g, '__escaped_single_quote__')
    .replace(/'/g, '"')
    .replace(/__escaped_single_quote__/g, '\\\'');

  // Remove string escapes except the ones followed by unicode u0027
  jsonString = jsonString.replace(/([^\\])\\(?!u0027)/g, '$1');

  // Wrap keys in double quotes
  jsonString = jsonString.replace(/[\s"]?([\w-]+)[\s"]?:/g, '"$1":');

  return JSON.parse(jsonString);
};
