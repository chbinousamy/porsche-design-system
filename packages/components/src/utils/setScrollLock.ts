export const setScrollLock = (isOpen: boolean): void => {
  const {
    body: { style, parentElement: htmlEl },
  } = document;

  if (isOpen) {
    style.top = -htmlEl.scrollTop + 'px';
    style.overflowY = 'scroll';
    style.position = 'fixed';
  } else {
    const topValue = -parseInt(style.top, 10);
    style.top = '';
    style.overflowY = '';
    style.position = '';
    htmlEl.scrollTop = topValue;
  }
};
