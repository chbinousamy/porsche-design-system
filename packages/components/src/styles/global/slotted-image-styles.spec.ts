import { componentFactory } from '../../test-utils';
import { expect } from '@jest/globals';
import * as slottedImageStylesUtils from './slotted-image-styles';
import { getSlottedImageStyles, tagNamesWithSlottedImageArray } from './slotted-image-styles';
import * as applyConstructableStylesheetStylesUtils from '../../utils/applyConstructableStylesheetStyle';

it.each(tagNamesWithSlottedImageArray)(
  'should return correct minified slotted image styles for component: %s',
  (tagName) => {
    expect(getSlottedImageStyles(tagName)).toMatchSnapshot();
  }
);

it.each(tagNamesWithSlottedImageArray)(
  'should apply constructable stylesheet with slotted image styles in connected callback for component %s',
  (tagName) => {
    const component = componentFactory(tagName);
    const getSlottedImageStylesSpy = jest.spyOn(slottedImageStylesUtils, 'getSlottedImageStyles');
    const applyConstructableStylesheetStylesSpy = jest.spyOn(
      applyConstructableStylesheetStylesUtils,
      'applyConstructableStylesheetStyles'
    );
    component.connectedCallback();

    expect(applyConstructableStylesheetStylesSpy).toHaveBeenCalledWith(component.host, getSlottedImageStyles);
    expect(getSlottedImageStylesSpy).toHaveBeenCalledWith(tagName);
  }
);
