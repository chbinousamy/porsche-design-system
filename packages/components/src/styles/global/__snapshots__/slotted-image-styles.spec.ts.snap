// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct minified slotted image styles for component: p-table 1`] = `
{
  "@global": {
    ":is(p-table) img": {
      "verticalAlign": "middle !important",
    },
  },
}
`;
